### ISSUE TITLE
Fix compilation errors for Oracle Packages, UDTs, ArrayRecords, Queues, Links in KotlinGenerator

### ISSUE THREAD

**Issue by davidkarlsen** [jOOQ/jOOQ#10778]
### Expected behavior
Should generate valid code.

### Actual behavior
```kotlin
/*
 * This file is generated by jOOQ.
 */
package com.edb.fs.tac.jfr.srv.jooq.packages


import com.edb.fs.tac.jfr.srv.jooq.Appdata
import com.edb.fs.tac.jfr.srv.jooq.packages.user_grant.Granttoroles

import org.jooq.Configuration
import org.jooq.impl.PackageImpl


/**
 * Convenience access to all stored procedures and functions in USER_GRANT
 */
@Suppress("UNCHECKED_CAST")
public class UserGrant extends PackageImpl {

    /**
     * The reference instance of <code>APPDATA.USER_GRANT</code>
     */
    val USER_GRANT = UserGrant()

    /**
     * Call <code>APPDATA.USER_GRANT.GRANTTOROLES</code>
     */
    fun granttoroles(
          configuration: Configuration
    ): Unit {
        val p = Granttoroles()

        p.execute(configuration)
    }

    /**
     * No further instances allowed
     */
    private UserGrant() {
        super("USER_GRANT", Appdata.APPDATA);
    }
}

```

```xml
<executions>
          <execution>
            <phase>generate-sources</phase>
            <configuration>
              <jdbc>
                <driver>${db.driver}</driver>
                <url>${db.url}</url>
                <user>${db.ddlusername}</user>
                <password>${db.ddlpassword}</password>
              </jdbc>
              <generator>
                <name>org.jooq.codegen.KotlinGenerator</name>
                <database>
                  <name>org.jooq.meta.oracle.OracleDatabase</name>
                  <inputSchema>appdata</inputSchema>
                  <recordVersionFields>VERSION</recordVersionFields>
                  <recordTimestampFields>UPDATED_DT</recordTimestampFields>
                  <excludes>
                    Batch.*
                    | Bin.*
                    | Flyway.*
                  </excludes>
                </database>
                <generate>
                  <javaTimeTypes>true</javaTimeTypes>
                  <jpaAnnotations>true</jpaAnnotations>
                  <jpaVersion>2.2</jpaVersion>
                  <pojosAsKotlinDataClasses>true</pojosAsKotlinDataClasses>
                  <pojosEqualsAndHashCode>true</pojosEqualsAndHashCode>
                  <pojos>true</pojos>
                  <daos>true</daos>
                  <validationAnnotations>false</validationAnnotations>
                  <springAnnotations>true</springAnnotations>
                </generate>
                <target>
                  <packageName>com.edb.fs.tac.jfr.srv.jooq</packageName>
                </target>
              </generator>
            </configuration>
            <goals>
              <goal>generate</goal>
            </goals>
          </execution>
```

### Versions

- jOOQ: 3.14.0
- Java: 15
- Database (include vendor): Oracle
- OS: N/A
- JDBC Driver (include name if inofficial driver): 19.3

**Comment by davidkarlsen** [comment]
Update: Still a problem with 3.14.1

**Comment by lukaseder** [comment]
Thanks for the update. Yes, this hasn't been fixed yet. The fix is more laborious than the ones that have been included in 3.14.1. We'll definitely look into this for 3.14.2

**Comment by lukaseder** [comment]
A list of limitations of the current `KotlinGenerator` includes:

- `org.jooq.Package`
- `org.jooq.UDT` (object types)
- `org.jooq.ArrayRecord` (table types)

**Comment by davidkarlsen** [comment]
Any eta on 3.14.2?

**Comment by lukaseder** [comment]
Anything you're looking for from that release, specifically? You can access snapshot builds from here, btw: https://www.jooq.org/download/versions

**Comment by lukaseder** [comment]
I'm currently working on a version of the generator that will make the compilation errors go away. That might not necessarily mean the object types are generated correctly, but I understand that the compilation errors cause some annoyance. It's possible, of course, to exclude the object types from being generated, see:
https://www.jooq.org/doc/latest/manual/code-generation/codegen-advanced/codegen-config-database/codegen-database-include-object-types/

**Comment by lukaseder** [comment]
Getting closer... There are still a few things that don't compile from our integration tests, but this should be doable for 3.14.2, which can be shipped next week.

**Comment by davidkarlsen** [comment]
Awesome. So excited about the Kotlin support.

**Comment by lukaseder** [comment]
Glad to hear that! :-) Please let us know if you'd like to see anything else in jOOQ's kotlin support, including improvements to the new `jOOQ-kotlin` module...

**Comment by lukaseder** [comment]
The compilation errors are now gone for our integration test database. There's still some work to do adding more integration tests, which can be done in later patch releases, though. Also, the known issues from the `JavaGenerator` (e.g. regarding overloading in some cases) may still be present.

I'll ship 3.14.2 including this fix (#10876) later today.

**Comment by lukaseder** [comment]
3.14.2 has shipped
